[feat/EATBOOK-39] 소설 상세 페이지 API 추가

cjm0611 opened the pull request:
## 설명
<!-- PR에서 해결하려는 문제나 기능을 간단히 설명합니다. -->
소설 상세 페이지에서 필요한 3가지 API를 구현했습니다.
- **소설 정보 조회 API**: 소설의 ID로 소설의 기본 정보(제목, 저자, 장르 등)를 조회합니다.
- **에피소드 목록 조회 API**: 소설의 에피소드 목록을 페이징하여 반환합니다.
- **가장 최근에 읽은 에피소드 조회 API**: 특정 사용자가 소설에서 마지막으로 읽은 에피소드 정보를 조회합니다.

## 포함된 변경 사항
<!-- 코드에서 어떤 변화가 이루어졌는지 요약합니다. 주요 변경 사항을 나열하세요. -->
- 위에서 서술한 API에 대한 컨트롤러, 서비스, 레포지터리, 테스트 코드를 추가했습니다.
- Comment, Favorite, ReadingLog Entity를 추가했습니다.
- QueryDSL을 도입했고, 이를 사용한 쿼리문은 Custom Repository를 생성하여 인터페이스와 구조체로 나눠 구현했습니다.
- checkStyle과 코드 포매터(Goole Style)를 도입하여 코드 스타일이 변경되었습니다.

## 테스트 방법
<!-- 변경된 부분을 테스트한 방법을 설명합니다. -->

- 테스트 코드 실행 (MemberServiceTest, NovelServiceTest, ReadingLogServiceTest)

## 추가 참고 사항
<!-- 리뷰어가 알아야 할 추가적인 정보나 참고해야 할 내용이 있다면 여기에 작성하세요. 예를 들어, 관련된 이슈 번호나 외부 링크 등을 포함할 수 있습니다. -->

- QueryDSL 관련 build.gradle은 [팀 노션 페이지](https://www.notion.so/cjm-0611/QueryDSL-1345b188f94a80cc87adc95e48e7469f?pvs=4)를 참고해주세요!
- 원래는 3가지 정보(소설 정보, 에피소드 목록, 독서기록)를 하나의 API로 구현하다가 분리한거라 하나의 branch가 되었는데, 변경된 파일 개수가 너무 많네요. 다음부터는 여러 개의 API를 구현할 때 PR을 각각 나눠서 올리겠습니다.

## 리뷰 요청 사항
<!-- 리뷰어가 특별히 검토해주었으면 하는 부분이 있다면 여기에 작성합니다. -->

- 쿼리문이 적절히 작성되었는지(JPA, QueryDSL), 메서드 네이밍이 괜찮은지, 테스트 코드는 적절히 작성되었는지 봐주시면 감사하겠습니다!

cjm0611 committed:
feat: 잘못된 쿼리 파라미터를 처리하는 global exception 추가

cjm0611 committed:
chore: querydsl 도입

cjm0611 committed:
feat: querydsl config 생성

cjm0611 committed:
fix: param 이름 명시적으로 수정

cjm0611 committed:
feat: 에피소드 정렬 순서 enum 생성

cjm0611 committed:
feat: 좋아요 Entity 생성

cjm0611 committed:
feat: 소설 도메인 예외처리 클래스 생성

cjm0611 committed:
feat: 소설 도메인 success response의 message, code를 관리하는 enum 생성

cjm0611 committed:
feat: 에피소드 Entity 출시일 관련 필드 네이밍 변경, 추가

cjm0611 committed:
chore: 에피소드 테이블 출시일 관련 column 변경으로 인한 config 최신화

cjm0611 committed:
fix: 레포지터리 Param 이름 명시적으로 변경

cjm0611 committed:
feat: 소설 좋아요 수, 특정 멤버의 좋아요 여부 가져오는 쿼리 메서드 구현

cjm0611 committed:
test: 소설 좋아요 수, 좋아요 여부가 맞는지 테스트하는 로직 추가

cjm0611 committed:
feat: Comment Entity 구현

cjm0611 committed:
feat: Comment 커스텀 레포지터리 구현

cjm0611 committed:
fix: LocalDateTime 타입에 length 제한 삭제

cjm0611 committed:
feat: 단일 소설 정보 GET API 구현

cjm0611 committed:
refactor: 북마크 누른 소설 리스트 받아오는 DTO 타입 수정

cjm0611 committed:
test: 북마크 누른 소설 리스트 DTO 타입 수정, 정적 팩토리 메서드 추가

cjm0611 committed:
chore: gitignore에 Q 객체 추가

cjm0611 committed:
feat: novelId로 소설 정보와 좋아요 수를 찾는 API 구현

cjm0611 committed:
test: 소설 서비스 테스트 코드 작성

소설 ID로 소설 객체를 찾는 method
소설 ID로 소설 정보와 좋아요 수를 조회하는 method

cjm0611 committed:
test: 소설 fixture 생성하는 코드 디렉토리 변경

cjm0611 committed:
style: 코드 포매터 적용

cjm0611 committed:
feat: 특정 소설에 대한 에피소드 목록을 조회하는 API 추가

cjm0611 committed:
fix: 공개 상태가 public인 에피소드만 가져오도록 수정

cjm0611 committed:
test: mock 함수 given 위치로 이동

cjm0611 committed:
fix: GlobalErrorMessage 별도의 파일로 분리

cjm0611 committed:
refactor: global한 success message 분리

cjm0611 committed:
fix: 북마크 누른 소설 필드, depth 수정

cjm0611 committed:
feat: 북마크를 누른 소설 API DTO 추가

cjm0611 committed:
feat: 소설의 에피소드 목록을 조회하는 API 구현

cjm0611 committed:
feat: DB 조회만 하는 서비스에 readOnly 어노테이션 추가

cjm0611 committed:
feat: 독서 기록 entity 추가

cjm0611 committed:
feat: 소설에서 가장 최근에 읽은 독서 기록을 조회하는 쿼리 구현

cjm0611 committed:
feat: 최근에 읽은 독서 기록을 조회하는 API 구현

cjm0611 committed:
docs: 독서 기록 JavaDoc 추가

cjm0611 committed:
test: 최근에 읽은 독서 기록을 조회하는 기능 테스트 작성

cjm0611 committed:
refactor: GlobalExceptionHandler method의 접근 제어자 변경

cjm0611 committed:
fix: 독서 기록 조회성공한 경우 적절한 message 반환

cjm0611 committed:
chore: gitignore에 env 파일 추가

cjm0611 committed:
chore: config 최신화

cjm0611 committed:
chore: QueryDSL 빌드 에러 수정

cjm0611 committed:
refactor: 경로 변수에 슬래시 포함

cjm0611 committed:
fix: 유저 에러 메세지 수정

pletain commented on src/main/java/com/ktb/eatbookappbackend/domain/member/controller/MemberController.java:
```diff
@@ -24,37 +29,33 @@ public class MemberController {
     @Authenticated
     @GetMapping("/bookmarks")
-    public ResponseEntity<?> getMemberBookmarkedNovels(@RequestParam final int page, @RequestParam final int size) {
+    public ResponseEntity<?> getMemberBookmarkedNovels(@RequestParam(name = "page") @Min(1) final int page,
```
유지보수 측면에서 와일드카드 대신 명확한 반환 타입을 지정하는 것이 더 좋을 거 같은데, 추후에 변경하실 생각이 있으신지 여쭤보고 싶습니다.

cjm0611 replied:
API의 성공, 실패 DTO 타입이 달라서 두 타입을 하나의 타입으로 표현하는 것이 어려워 와일드카드를 사용했습니다. 두 타입을 하나의 타입으로 어떻게 표현하면 될까요? ResponseEntity<Object>로 표현하는 정도로 괜찮을까요?

pletain replied:
ResponseEntity로 표현하는 것도 동일하게 명확한 반환 타입이 없다고 생각됩니다. 저는 아래 두 방식이 떠올라서 제안드려봅니다!

1. 두 DTO가 상속하는 공통 인터페이스나 Wrapper클래스를 만들어서, 이를 반환타입으로 사용
2. 추후에 공통 예외 처리 로직을 도입해서, 컨트롤러에서는 성공 DTO을 반환타입으로 사용하고, 공통 예외처리에선 실패 DTO를 반환타입으로 사용

cjm0611 replied:
이전에 달았던 코멘트에 오타가 났네요. `ResponseEntity<Object>` 정도로 표현해도 괜찮은지 여쭤보려고 했습니다.
제안 주신 방법 중 2번 방법을 MVP 구현 이후에 도입하는 방안으로 진행하겠습니다! 백엔드 노션 문서에 리팩터링 항목으로 적어뒀습니다.

pletain commented on src/main/java/com/ktb/eatbookappbackend/domain/member/controller/MemberController.java:
```diff
@@ -24,37 +29,33 @@ public class MemberController {
-        PaginationWithDataDTO<MemberBookmarkedNovelDTO> bookmarkedNovels = memberService.getMemberBookmarkedNovels(memberId, page, size);
-
+        BookmarkedNovelsPaginationDTO bookmarkedNovels = memberService.getMemberBookmarkedNovels(memberId, page, size);
         if (page - 1 > bookmarkedNovels.pagination().totalPages()) {
```
1. 이 부분은 MemberService에서 처리하면 어떨까 제안드립니다! 구체적으론 Service에서 예외를 던지면 이를 캐치해서 컨트롤러나 전역 예외처리에서 처리하는게 좋을거 같습니다.
Controller는 요청을 받고, Service 계층에서 비즈니스 로직을 처리 후 그 결과를 반환하는 역할을 합니다.
하지만 이 부분은 비즈니스 로직에 가까운 검증 로직이라 생각됩니다. 단순히 입력 값의 형식이나 범위를 확인하는게 아니라, 데이터베이스 조회 결과를 기반으로 한 로직이기 때문입니다. 

2. 이 비교문의 형태가 전체 페이지가 0부터 시작하기 때문에라면,
page - 1 > bookmarkedNovels.pagination().totalPages() 대신 page > bookmarkedNovels.pagination().totalPages() + 1로 수정하는 편이 좀 더 의도가 명확할 거 같아서 제안드려봅니다!

pletain commented on src/main/java/com/ktb/eatbookappbackend/domain/readingLog/controller/ReadingLogController.java:
```diff
@@ -0,0 +1,50 @@
+            return ResponseEntity.ok(SuccessResponseDTO.of(GlobalSuccessMessage.NO_RESULTS_FOUND, null));
+        }
+
+        Map<String, Object> data = new HashMap<>();
```
명확한 데이터 구조를 위해서 Map 대신 별도의 응답 DTO를 사용하는 것을 제안 드립니다!  타입 안정성과 가독성 측면에서도 더 개선될거 같습니다

pletain commented on src/main/java/com/ktb/eatbookappbackend/domain/readingLog/service/ReadingLogService.java:
```diff
@@ -0,0 +1,35 @@
+     * @return 마지막으로 읽은 에피소드에 대한 {@link ContinueReadingInfoDTO}를 포함하는 {@link Optional}.
+     *         마지막으로 읽은 에피소드가 없는 경우, {@link Optional}은 비어 있습니다.
+     */
+    @Transactional(readOnly = true)
```
조회 메서드에 `@Transactional(readOnly = true)`를 설정해서 데이터 일관성 유지하면서 메모리도 절약할 수 있어서 좋은 부분이라고 생각합니다!

pletain commented on src/main/java/com/ktb/eatbookappbackend/domain/readingLog/service/ReadingLogService.java:
```diff
@@ -0,0 +1,35 @@
+     *         마지막으로 읽은 에피소드가 없는 경우, {@link Optional}은 비어 있습니다.
+     */
+    @Transactional(readOnly = true)
+    public Optional<ContinueReadingInfoDTO> getLastReadEpisode(String memberId, String novelId) {
```
현재 코드에서 Optional을 사용하고 있는데, 불필요한 Optional 사용은 메모리 오버헤드가 발생할 수 있고, 컨트롤러에서 이를 처리하는 로직이 복잡해질 수 있다고 생각합니다. 예) `lastReadEpisodeOpt.get()`사용으로 인해 추가적인 검사가 필요
따라서 Service 계층에서 Optional 대신 비어 있는 DTO 객체나 null을 반환하여 컨트롤러에서 단순한 null 체크만으로 처리할 수 있도록 변경하는 것이 어떤지 제안드립니다!

[참고자료1](https://velog.io/@hope1213/Optional%EC%9D%80-%EC%99%9C-%EC%82%AC%EC%9A%A9%ED%95%98%EB%8A%94%EC%A7%80-%EC%82%AC%EC%9A%A9%EC%8B%9C-%EC%A3%BC%EC%9D%98%EC%82%AC%ED%95%AD) [참고자료2](https://dzone.com/articles/using-optional-correctly-is-not-optional) [참고자료3](https://dev-coco.tistory.com/178)

cjm0611 replied:
Optional을 사용하면 서비스 계층에서 `값이 없을 수 있다`는 의미가 명시적으로 표현할 수 있습니다. Optional을 사용하지 않는다면, 컨트롤러에서 각 DTO나 객체에 대해 null 가능성을 염두에 두고 검사를 해야 합니다. 하지만 Optional 키워드를 일관적으로 사용하면 Optional 키워드가 있는 곳에서만 null 여부를 확인하면 되므로 가독성과 유지보수성에 좋다고 생각합니다.
결국 `메모리 오버헤드를 줄이고 코드의 간결함을 위한 null 사용`과 `Optional을 통해 명시적으로 빈 상태를 표현하여 유지보수성 향상` 사이의 trade-off로 볼 수 있을 것 같은데, 저는 불필요한 Optional 사용은 아니라고 생각합니다!

pletain commented on src/main/java/com/ktb/eatbookappbackend/domain/member/service/MemberService.java:
```diff
@@ -1,47 +1,55 @@
+    public BookmarkedNovelsPaginationDTO getMemberBookmarkedNovels(String memberId, int page, int size) {
         PageRequest pageRequest = PageRequest.of(page - 1, size);
         Page<Bookmark> bookmarkPage = bookmarkRepository.findByMemberIdWithNovel(memberId, pageRequest);
 
-        PaginationInfoDTO paginationInfo = new PaginationInfoDTO(
-                page,
-                size,
-                (int) bookmarkPage.getTotalElements(),
-                bookmarkPage.getTotalPages()
+        PaginationInfoDTO paginationInfo = PaginationInfoDTO.of(
+            page,
+            size,
+            (int) bookmarkPage.getTotalElements(),
+            bookmarkPage.getTotalPages()
         );
 
-        List<MemberBookmarkedNovelDTO> bookmarks = bookmarkPage.getContent().stream()
-                .map(bookmark -> MemberBookmarkedNovelDTO.of(bookmark.getNovel()))
-                .collect(Collectors.toList());
+        List<NovelDTO> bookmarks = bookmarkPage.getContent().stream()
+            .map(bookmark -> {
+                Novel novel = bookmark.getNovel();
+                int favoriteCount = favoriteRepository.countByNovelId(novel.getId());
+                return NovelDTO.of(bookmark.getNovel(), favoriteCount);
+            })
+            .collect(Collectors.toList());
 
```
현재 방식은 Bookmark 조회 1번, Favorite 조회 N번(bookmark된 소설 수만큼) 데이터베이스에 호출하는거 같은데,
이 부분을 아래와 같은 하나의 쿼리를 사용해 개선할 수도 있을거 같습니다.
```
    @Query("SELECT b, COUNT(f) AS favoriteCount " +
            "FROM Bookmark b " +
            "LEFT JOIN Favorite f ON f.novel.id = b.novel.id " +
            "WHERE b.member.id = :memberId " +
            "GROUP BY b")
    BookmarkWithFavoriteCountDTO findBookmarksByMemberIdWithFavoriteCount(@Param("memberId") String memberId);
```
하지만 쿼리가 복잡해서, novel_id, member_id 등에 인덱스가 적절히 설정되지 않으면 JOIN, GROUP BY 연산이 느려질 수 있을거 같습니다. 또한 데이터의 양에 따라서 그 결과값도 큰 차이가 있을거 같습니다.
결론적으로 추후에 이 부분은 같이 테스트 및 실험해보면서 리팩토링하면 더 좋을거 같습니다!

cjm0611 replied:
제안주신 쿼리에 페이지네이션 적용이 안되어 있는 것 같습니다. 페이지네이션 적용하여 변경하면 기존보다 성능이 좋아질 것은 분명해보이는데, 데이터의 양에 따라서 얼마나 성능이 개선될지 차이가 날 것 같아서 말씀주신 것처럼 추후에 리팩터링해보면 좋을 것 같습니다! 이것도 백엔드 문서에 적어뒀습니다.

cjm0611 committed:
feat: global 예외와 핸들러 추가

cjm0611 committed:
refactor: page 유효성 검사 서비스로 이동

cjm0611 committed:
feat: 성공 response dto의 data type 변경

cjm0611 committed:
refactor: API 반환 타입 명시

HashMap으로 표현하던 response data의 타입 명시
변경은 특정 회원이 북마크한 소설 목록을 페이지네이션하여 조회하는 API에서 발생

cjm0611 committed:
fix: ReadingLog의 ttsLastPositionSeconds type Time으로 변경

cjm0611 committed:
refactor: 최근 읽은 독서기록 DTO 명시적 반환
